File: calc.py
#!/usr/bin/python3

NUM_STEPS = 1024;

WIDTH = 400
HEIGHT = 400

def __scale(color: tuple, a: float) -> tuple:
    assert len(color) == 3
    return (color[0] * a, color[1] * a, color[2] * a)

# tmp solution
COLORS = (__scale((0.61, 0.15, 0.69), 255),
          __scale((0.25, 0.32, 0.71), 255),
          __scale((0.00, 0.66, 0.96), 255))

def __color_sum(color1: tuple, color2: tuple) -> tuple:
    assert len(color1) == 3
    assert len(color2) == 3
    return (color1[0] + color2[0], color1[1] + color2[1], color1[2] + color2[2])

def _get_color(n: int, colors: tuple) -> tuple:
    if (n == NUM_STEPS):
        return (0.0 ,0.0 ,0.0)
    
    i = n % 20 / 20
    cel = n // 20 % 3

    color1 = colors[0]
    color2 = colors[1]
    color3 = colors[2]

    if (cel == 0):
        return __color_sum(__scale(color1, i), __scale(color2, (1 - i)))

    if (cel == 1):
        return __color_sum(__scale(color3, i), __scale(color1, (1 - i)))

    if (cel == 2):
        return __color_sum(__scale(color2, i), __scale(color3, (1 - i)))

    return (0.0, 0.0, 0.0)

def _julia_number_of_term(x: float, y: float, cx: float, cy: float) -> int:
    x0, y0 = x, y
    i = 0
    while (x0 * x0 + y0 * y0 <= 4 and i < NUM_STEPS):
        x0, y0 = x0 * x0 - y0 * y0 + cx, 2 * x0 * y0 + cy
        i += 1

    return i

def _mandelbrot_number_of_term(x: float, y: float) -> int:
    return _julia_number_of_term(0, 0, x, y)

def calc(mx: float, my: float, dmx: float, dmy: float,
        cx: float, cy: float,
        jx: float, jy: float, djx: float, djy: float) -> tuple:
    """
    This is the main function of module. It calculates the colors of pixels
    for requested areas of Mandelbrot and Julia sets.

    @param mx --- real part of center of requested area of Mandelbrot set
    @param my --- imaginary part of center of requested area of Mandelbrot set
    @param dmx --- width (size along real axis) of requested area of Mandelbrot set
    @param dmy --- height (size along imaginary axis) of requested area of Mandelbrot set

    @param cx --- real coord of requested c-point for Julia set
    @param cy --- imaginary coord of requested c-point for Julia set

    @param jx --- real part of center of requested area of Julia set
    @param jy --- imaginary part of center of requested area of Julia set
    @param djx --- width (size along real axis) of requested area of Julia set
    @param djy --- height (size along imaginary axis) of requested area of Julia set

    @return typle --- 2d list of pixel colors (color --- rgb tuple)
    """
    m_points = [[(0, 0, 0) for _ in range(HEIGHT)] for _ in range(WIDTH)]
    for i in range(WIDTH):
        for j in range(HEIGHT):
            m_points[i][j] = _get_color(_mandelbrot_number_of_term(
                mx - dmx / 2 + dmx * i / HEIGHT, my - dmy / 2 + dmy * j / HEIGHT), COLORS)
    j_points = [[(0, 0, 0) for _ in range(HEIGHT)] for _ in range(WIDTH)]
    for i in range(WIDTH):
        for j in range(HEIGHT):
            j_points[i][j] = _get_color(_julia_number_of_term(
                jx - djx / 2 + djx * i / HEIGHT, jy - djy / 2 + djy * j / HEIGHT, cx, cy), COLORS)
    return (m_points, j_points)

if (__name__ == "__main__"):
    print("TODO: write test script")

File: calc.c
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>

const int NUM_STEPS = 1024;

static int col1[] = {156,  38, 177};
static int col2[] = { 64,  82, 182};
static int col3[] = {  0, 169, 246};

int _set_color(int r1, int g1, int b1,
              int r2, int g2, int b2,
              int r3, int g3, int b3) {
    col1[0] = r1;
    col1[1] = g1;
    col1[2] = b1;
    col2[0] = r2;
    col2[1] = g2;
    col2[2] = b2;
    col3[0] = r3;
    col3[1] = g3;
    col3[2] = b3;

    return 0;
}

static void *__vec3(int r, int g, int b) {
    int *ans = malloc(sizeof(int) * 3);
    *ans = r;
    *(ans+1) = g;
    *(ans+2) = b;
    return ans;
}

static void *__mix(double a1, double a2, double a3) {
    return __vec3(col1[0] * a1 + col2[0] * a2 + col3[0] * a3,
                col1[1] * a1 + col2[1] * a2 + col3[1] * a3,
                col1[2] * a1 + col2[2] * a2 + col3[2] * a3);
}

static void *__constructColor(int iter) {
    if (iter == NUM_STEPS) {
        return __vec3(0, 0, 0);
    }

    double i = iter % 20 / 20.0;
    int cel = iter / 20 % 3;

    if (cel == 0) {
        return __mix(i, 1 - i, 0);
    }
    if (cel == 1) {
        return __mix(1 - i, 0, i);
    }
    if (cel == 2) {
        return __mix(0, i, 1 - i);
    }
    return __vec3(0, 0, 0);
}

void *_calc_man(int width, int height,
        double mx, double my, double dmx, double dmy) {

    int *ans = malloc(sizeof(int)*width*height*3);
    double x, y, x0, y0, tmp;
    int *col;
    int cnt;
    for (int i = 0; i < width; ++i) {
        for (int j = 0; j < height; ++j) {
            x0 = 0;
            y0 = 0;
            x = mx - dmx / 2 + dmx * i / width;
            y = my - dmy / 2 + dmy * j / height;
            cnt = 0;
            while (x0 * x0 + y0 * y0 < 4 && cnt < NUM_STEPS) {
                tmp = x0 * x0 - y0 * y0 + x;
                y0 = 2 * x0 * y0 + y;
                x0 = tmp;
                cnt++;
            }
            col = (int *)__constructColor(cnt);
            *(ans + height * 3 * i + j * 3 + 0) = col[0];
            *(ans + height * 3 * i + j * 3 + 1) = col[1];
            *(ans + height * 3 * i + j * 3 + 2) = col[2];
            free(col);
        }
    }
    return ans;
}

void *_calc_jul(int width, int height,
        double cx, double cy,
        double jx, double jy, double djx, double djy) {

    int *ans = malloc(sizeof(int)*width*height*3);
    double x0, y0, tmp;
    int *col;
    int cnt;
    for (int i = 0; i < width; ++i) {
        for (int j = 0; j < height; ++j) {
            x0 = jx - djx / 2 + djx * i / width;
            y0 = jy - djy / 2 + djy * j / height;
            cnt = 0;
            while (x0 * x0 + y0 * y0 < 4 && cnt < NUM_STEPS) {
                tmp = x0 * x0 - y0 * y0 + cx;
                y0 = 2 * x0 * y0 + cy;
                x0 = tmp;
                cnt++;
            }
            col = (int *)__constructColor(cnt);
            *(ans + height * 3 * i + j * 3 + 0) = col[0];
            *(ans + height * 3 * i + j * 3 + 1) = col[1];
            *(ans + height * 3 * i + j * 3 + 2) = col[2];
            free(col);
        }
    }
    return ans;
}

void _free(void * ptr) {
    free(ptr);
}
